fem/libceed/diffusion.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "ceed.hpp"
+#include "../../general/device.hpp"
+
+#include "diffusion.h"
+
+namespace mfem
+{
+
+#ifdef MFEM_USE_CEED
+
```

Comment:
Why are there separate codes (in diffusion.cpp and mass.cpp) enclosing the kernels for mass and diffusion? I would expect computing some values at quad points to be conceptually the same, up to the kernels. The two files mass.cpp and diffusion.cpp two are nearly identical - can't they be combined?

If not, do you expect app developers to write a function like this for every custom integrator, only to compute the quad data? And this doesn't include the actual kernels.

Reply:
It's true that these codes are largely similar, I was planning on factoring them in the future, when I would write a PR for vectorial mass and diffusion.